`pico-blink` Build Error: Undefined symbols for architecture armv6m:   "___muldi3"

When trying to build the `pico-blink` example on macOS with:

```
cd swift-embedded-examples/pico-blink
rm -rf .build ; TOOLCHAINS=org.swift.59202404131a ./build.sh
```

I get the following error:

```
# Link
$CLANG .build/release/Support.build/{Support.c,crt0.S}.o .build/release/Blinky.build/*.o -target armv6m-apple-none-macho -o $BUILDROOT/blinky $LD_FLAGS
+ /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2024-04-13-a.xctoolchain/usr/bin/clang .build/release/Support.build/Support.c.o .build/release/Support.build/crt0.S.o .build/release/Blinky.build/Blinky.swift.o -target armv6m-apple-none-macho -o /path/to/swift-embedded-examples/pico-blink/.build/armv6m-apple-none-macho/release/blinky -static -Wl,-e,_reset -dead_strip -Wl,-no_zero_fill_sections -Wl,-segalign,4 -Wl,-segaddr,__RESET,0x20000000 -Wl,-segaddr,__VECTORS,0x20000100 -Wl,-seg1addr,0x20000200 -Wl,-pagezero_size,0 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/15.0.0/lib/darwin/macho_embedded/libclang_rt.soft_static.a
ld: warning: ignoring file /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2024-04-13-a.xctoolchain/usr/lib/clang/17/lib/darwin/macho_embedded/libclang_rt.soft_static.a, building for free standing-armv6m but attempting to link with file built for macOS-armv7
ld: warning: ignoring file /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/15.0.0/lib/darwin/macho_embedded/libclang_rt.soft_static.a, building for free standing-armv6m but attempting to link with file built for macOS-armv7
Undefined symbols for architecture armv6m:
  "___muldi3", referenced from:
      _$s6Blinky4MainV5delayyySiFZTf4nd_n in Blinky.swift.o
ld: symbol(s) not found for architecture armv6m
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Is the `___muldi3` symbol not available on armv6m? If not, why is it being used?

System information:

```
swift-DEVELOPMENT-SNAPSHOT-2024-04-13-a.xctoolchain
Apple Swift version 6.0-dev (LLVM 01231f56c62f4c4, Swift abaa63821080b92)
Darwin Kernel Version 23.4.0: Fri Mar 15 00:12:25 PDT 2024; root:xnu-10063.101.17~1
```

Toolchain ID:
```
$ plutil -extract CFBundleIdentifier raw -o - /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-*.xctoolchain/Info.plist                                    
org.swift.59202404131a
```

Git commit of `swift-embedded-examples`: 
```
69dc23e718c5b20797daec17d4754cd42bb21c24
```

Full build log: [pico-blink-build-log.txt](https://github.com/apple/swift-embedded-examples/files/15029563/pico-blink-build-log.txt)

I can sort-of reproduce the problem. One issue is that this part of the build script for pico-blink should not be necessary anymore, as the expectation is that the downloadable swift.org toolchain includes a valid libclang_rt.soft_static.a for the CPU arch we're building here (armv6m):

```
XCODE_RESOURCE_DIR=`env TOOLCHAINS="" xcrun clang -print-resource-dir`
LD_FLAGS+=" $XCODE_RESOURCE_DIR/lib/darwin/macho_embedded/libclang_rt.soft_static.a"
```

Once I drop this, I get:

```
# Link
$CLANG .build/release/Support.build/{Support.c,crt0.S}.o .build/release/Blinky.build/*.o -target armv6m-apple-none-macho -o $BUILDROOT/blinky $LD_FLAGS
+ /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2024-04-25-a.xctoolchain/usr/bin/clang .build/release/Support.build/Support.c.o .build/release/Support.build/crt0.S.o .build/release/Blinky.build/Blinky.swift.o -target armv6m-apple-none-macho -o /Users/kuba/Documents/swift-embedded-examples/pico-blink/.build/armv6m-apple-none-macho/release/blinky -static -Wl,-e,_reset -dead_strip -Wl,-no_zero_fill_sections -Wl,-segalign,4 -Wl,-segaddr,__RESET,0x20000000 -Wl,-segaddr,__VECTORS,0x20000100 -Wl,-seg1addr,0x20000200 -Wl,-pagezero_size,0
ld: warning: ignoring file /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2024-04-25-a.xctoolchain/usr/lib/clang/17/lib/darwin/macho_embedded/libclang_rt.soft_static.a, building for free standing-armv6m but attempting to link with file built for macOS-armv7
Undefined symbols for architecture armv6m:
  "___muldi3", referenced from:
      _$s6Blinky4MainV5delayyySiFZTf4nd_n in Blinky.swift.o
ld: symbol(s) not found for architecture armv6m
```

The problem is that libclang_rt.soft_static.a is somehow only armv7 and contains no other slices. That will need to be fixed, but I'm not sure how that happens in the first place.

The only workaround I can think of (short of actually removing the multiplication from source code, but sometimes those are really useful 😄) would be to rebuild libclang_rt from LLVM's compiler-rt from source and explicitly build the armv6m slice, and link that instead.

I'll use this issue to track fixing the `Undefined symbols for architecture armv6m: "___muldi3"` problem, for other issues please file a separate problem report.

--- a/utils/swift_build_support/swift_build_support/products/llvm.py
+++ b/utils/swift_build_support/swift_build_support/products/llvm.py
@@ -142,7 +142,8 @@ class LLVM(cmake_product.CMakeProduct):
                     if not os.path.isfile(dest_lib_path):
                         if os.path.isfile(host_lib_path):
                             if _os == 'tvos':
-                                print('{} -> {} (stripping i386)'.format(host_lib_path, dest_lib_path))
+                                print('{} -> {} (stripping i386)'.format(
+                                    host_lib_path, dest_lib_path))
                                 self.copy_lib_stripping_architecture(host_lib_path,
                                                                      dest_lib_path,
                                                                      'i386')
@@ -163,11 +164,13 @@ class LLVM(cmake_product.CMakeProduct):
                             if _os == 'tvos':
                                 # This is to avoid strip failures when generating
                                 # a toolchain
-                                print('{} -> {} (stripping i386)'.format(host_sim_lib_path, dest_sim_lib_path))
+                                print('{} -> {} (stripping i386)'.format(
+                                    host_sim_lib_path, dest_sim_lib_path))
                                 self.copy_lib_stripping_architecture(
                                     host_sim_lib_path, dest_sim_lib_path, 'i386')
                             else:
-                                print('{} -> {}'.format(host_sim_lib_path, dest_sim_lib_path))
+                                print('{} -> {}'.format(
+                                    host_sim_lib_path, dest_sim_lib_path))
                                 shutil.copy(host_sim_lib_path, dest_sim_lib_path)
 
                         elif os.path.isfile(host_lib_path):
@@ -185,7 +188,8 @@ class LLVM(cmake_product.CMakeProduct):
                                 shell.call(['lipo', '-remove', 'i386', host_lib_path,
                                             '-output', dest_sim_lib_path])
                             else:
-                                print('{} -> {}'.format(host_lib_path, dest_sim_lib_path))
+                                print('{} -> {}'.format(host_lib_path,
+                                                        dest_sim_lib_path))
                                 shutil.copy(host_lib_path, dest_sim_lib_path)
                         elif self.args.verbose_build:
                             print('no file exists at {}', host_sim_lib_path)
@@ -294,8 +298,9 @@ class LLVM(cmake_product.CMakeProduct):
 
         if system() == 'Darwin':
             # Ask for Mach-O cross-compilation builtins (for Embedded Swift)
-            llvm_cmake_options.define('COMPILER_RT_FORCE_BUILD_BAREMETAL_MACHO_BUILTINS_ARCHS:STRING',
-                                      'armv6 armv6m armv7 armv7m armv7em')
+            llvm_cmake_options.define(
+                'COMPILER_RT_FORCE_BUILD_BAREMETAL_MACHO_BUILTINS_ARCHS:STRING',
+                'armv6 armv6m armv7 armv7m armv7em')
 
         llvm_enable_projects = ['clang']
 

--- a/utils/swift_build_support/swift_build_support/products/llvm.py
+++ b/utils/swift_build_support/swift_build_support/products/llvm.py
@@ -142,10 +142,12 @@ class LLVM(cmake_product.CMakeProduct):
                     if not os.path.isfile(dest_lib_path):
                         if os.path.isfile(host_lib_path):
                             if _os == 'tvos':
+                                print('{} -> {} (stripping i386)'.format(host_lib_path, dest_lib_path))
                                 self.copy_lib_stripping_architecture(host_lib_path,
                                                                      dest_lib_path,
                                                                      'i386')
                             else:
+                                print('{} -> {}'.format(host_lib_path, dest_lib_path))
                                 shutil.copy(host_lib_path, dest_lib_path)
                         elif self.args.verbose_build:
                             print('no file exists at {}'.format(host_lib_path))
@@ -161,9 +163,11 @@ class LLVM(cmake_product.CMakeProduct):
                             if _os == 'tvos':
                                 # This is to avoid strip failures when generating
                                 # a toolchain
+                                print('{} -> {} (stripping i386)'.format(host_sim_lib_path, dest_sim_lib_path))
                                 self.copy_lib_stripping_architecture(
                                     host_sim_lib_path, dest_sim_lib_path, 'i386')
                             else:
+                                print('{} -> {}'.format(host_sim_lib_path, dest_sim_lib_path))
                                 shutil.copy(host_sim_lib_path, dest_sim_lib_path)
 
                         elif os.path.isfile(host_lib_path):
@@ -181,6 +185,7 @@ class LLVM(cmake_product.CMakeProduct):
                                 shell.call(['lipo', '-remove', 'i386', host_lib_path,
                                             '-output', dest_sim_lib_path])
                             else:
+                                print('{} -> {}'.format(host_lib_path, dest_sim_lib_path))
                                 shutil.copy(host_lib_path, dest_sim_lib_path)
                         elif self.args.verbose_build:
                             print('no file exists at {}', host_sim_lib_path)
@@ -195,6 +200,7 @@ class LLVM(cmake_product.CMakeProduct):
                     dest_lib_path = os.path.join(dest_builtins_dir, lib_name)
                     if not os.path.isfile(dest_lib_path):
                         if os.path.isfile(host_lib_path):
+                            print('{} -> {}'.format(host_lib_path, dest_lib_path))
                             shutil.copy(host_lib_path, dest_lib_path)
                         elif self.args.verbose_build:
                             print('no file exists at {}'.format(host_lib_path))
@@ -285,6 +291,11 @@ class LLVM(cmake_product.CMakeProduct):
         llvm_cmake_options.define('LLVM_INCLUDE_DOCS:BOOL', 'TRUE')
         llvm_cmake_options.define('LLVM_ENABLE_LTO:STRING', self.args.lto_type)
         llvm_cmake_options.define('COMPILER_RT_INTERCEPT_LIBDISPATCH', 'ON')
+
+        if system() == 'Darwin':
+            # Ask for Mach-O cross-compilation builtins (for Embedded Swift)
+            llvm_cmake_options.define('COMPILER_RT_FORCE_BUILD_BAREMETAL_MACHO_BUILTINS_ARCHS:STRING',
+                                      'armv6 armv6m armv7 armv7m armv7em')
 
         llvm_enable_projects = ['clang']
 

